plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/weighers/KindWeigher.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+package org.jetbrains.kotlin.idea.completion.weighers
+
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementWeigher
+import com.intellij.openapi.util.Key
+import org.jetbrains.kotlin.idea.completion.KeywordLookupObject
+import org.jetbrains.kotlin.idea.completion.contributors.keywords.ReturnKeywordHandler.isReturnAtHighlyLikelyPosition
+import org.jetbrains.kotlin.idea.completion.lookups.KotlinCallableLookupObject
+import org.jetbrains.kotlin.idea.completion.lookups.factories.PackagePartLookupObject
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtEnumEntrySymbol
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtSymbol
+import org.jetbrains.kotlin.psi.UserDataProperty
+
+internal object KindWeigher {
+    private enum class Weight {
+        RETURN_AT_HIGHLY_LIKELY_POSITION,
+        ENUM_MEMBER,
+        CALLABLE,
+        KEYWORD,
+        DEFAULT,
+        PACKAGES
+    }
+
+    const val WEIGHER_ID = "kotlin.kind"
+
+    private var LookupElement.isEnumEntry by UserDataProperty(Key<Boolean>("KOTLIN_KIND_WEIGHER_IS_ENUM"))
+
+    fun KtAnalysisSession.addWeight(lookupElement: LookupElement, symbol: KtSymbol) {
+        if (symbol is KtEnumEntrySymbol) {
+            lookupElement.isEnumEntry = true
+            return
+        }
+    }
+
+    object Weigher : LookupElementWeigher(WEIGHER_ID) {
+        override fun weigh(element: LookupElement): Comparable<Nothing> {
+            if (element.isEnumEntry == true) return Weight.ENUM_MEMBER
+            if (element.isReturnAtHighlyLikelyPosition == true) return Weight.RETURN_AT_HIGHLY_LIKELY_POSITION
```

Comment:
It is a little bit confusing to me that we have this category (RETURN_AT_HIGHLY_LIKELY_POSITION) right along with the types of the expressions (Keywords, Enums, etc.)

Maybe we should rename something?

Reply:
It's called `probableKeyword` at in FE1.0 (https://github.com/JetBrains/intellij-community/blob/390aca4aebb66c0b235b1515a8228d2deedf0f6c/plugins/kotlin/completion/src/org/jetbrains/kotlin/idea/completion/Weighers.kt#L106). I wanted to pick a more descriptive name. How about `RETURN_KEYWORD_AT_HIGHLY_LIKELY_POSITION`?